agent/react_graph.py

```diff
@@ -169,25 +189,45 @@ async def acall_model(state: UserState, config: RunnableConfig):
         res = await model_runnable.ainvoke({"messages": messages}, config)
         return {"messages": [res]}
 
+    def request_login_node(_: UserState):
+        """
+        If the user needs to log in, this node sends a message to the user.
+        """
+        return {
+            "messages": [
+                AIMessage(
+                    content="This action requires you to be signed in. Please log in and then try again."
+                )
+            ]
+        }
+
     def agent_should_continue(
-        state: UserState,
-    ) -> Literal["booking_validation", "continue", "end"]:
+        state: UserState, config: RunnableConfig
+    ) -> Literal["booking_validation", "continue", "request_login", "end"]:
         """
         Function to determine which node is called after the agent node.
         """
         messages = state["messages"]
         last_message = messages[-1]
-        # If the LLM makes a tool call, then we route to the "tools" node
-        if hasattr(last_message, "tool_calls") and len(last_message.tool_calls) > 0:
-            confirmation_needing_tools = get_confirmation_needing_tools()
-            for tool_call in last_message.tool_calls:
-                tool_name = tool_call["name"]
-                if tool_name in confirmation_needing_tools:
-                    if tool_name == "insert_ticket":
-                        return "booking_validation"
-            return "continue"
-        # Otherwise, we stop (reply to the user)
-        return "end"
+
+        # First check if the last message has tool calls.
+        if not hasattr(last_message, "tool_calls") or len(last_message.tool_calls) == 0:
+            return "end"
+
+        # Next, check if any tool requires authentication.
+        for tool_call in last_message.tool_calls:
+            if tool_call["name"] in get_auth_tools():
+                if not __is_logged_in(config):
+                    return "request_login"
+
+        # If authentication passes, then check if any tool needs user confirmation.
+        for tool_call in last_message.tool_calls:
```

Comment:
```suggestion
                    return "request_login"
```

Might be able to check both in the same for loop?